### PR TITLE
feat(remix-react): add support for `http-equiv` in `meta` function

### LINKED ISSUE #941: [Feature]: Support meta http-equiv tag
### What is the new or updated feature that you are suggesting?

I would for like Remix to allow me to add a [`<meta http-equiv>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta#attr-http-equiv) tag to my page via the meta export.

### Why should this feature be included?

I would like for the meta export to support `http-equiv` because there doesn't appear to be any other way to add a `<meta http-equiv>` tag a single route.

If this change sounds good, I'm happy to put together a PR adding a special case for http-equiv. I could add [`charset`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta#attr-charset) too for good measure if that's desired.

### PR DESCRIPTION
**Pull request by MichaelDeBoey**
Took inspiration of the OG tags: if people want to use the `http-equiv` tag, they'll have to do

```ts
export const meta: MetaFunction = () => ({
  'http-equiv:refresh': '3;url=https://www.mozilla.org',
});
```

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta#attr-http-equiv

@kentcdodds Not sure how this would best fit in the docs, so I'll leave that up to you if you don't mind.

---

Closes #941

### CODE REVIEW

**Comment by kentcdodds** [comment]
I wonder... Are there a lot of unique tags like this? Should we maybe allow people to have a different way to define these for more control?

**Comment by MichaelDeBoey** [comment]
It seems like those are the only ones

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta

**Review by kentcdodds — APPROVED** [review]
Looks good to me 👍

**Comment by kiliman** [comment]
Remix currently special-cases `og:*` prefix to set the `property` attribute vs `name`.

However, `fb:*` prefix also uses `property`, but `twitter:*` uses `name`. Who needs consistency?! 🤦‍♂️

I discuss in this issue. https://github.com/remix-run/remix/issues/1717

**Comment by machour** [comment]
@kentcdodds it's not so meaningless, as it allows to override a specific item in the merged metas:

// parent.tsx
```
export const meta: MetaFunction = () => {
  return {
    author1: {name: "author", content: "Author One"},
    author2: {name: "author", content: "Author Two"},
  };
};
```

// parent/child.tsx

```
export const meta: MetaFunction = () => {
  return {
    author2: {name: "author", content: "Another Author"},
  };
};
```

in the child route, this will be rendered: 

```
<meta name="author" content="Author One" />
<meta name="author" content="Another Author" /> <-- took the place of Author Two
```

(the specification allow some meta to appear more than once)

**Comment by kentcdodds** [comment]
Oh nice! Yeah, I'm a fan 👍

**Review by kentcdodds — APPROVED** [review]
Love it 👍

Would you mind adding some docs for this?

File: packages/remix-react/components.tsx
Comment (machour) [inline packages/remix-react/components.tsx:687]: We still need keys on all elements, as we're in a loop.

Just a question on code readability: wouldn't it be better to break that giant ternary operation into something like:

```js
if (["charset", "charSet"].includes(name)) {
  return <meta key="charset" charSet={value} />
}

if (name === "title") { 
 // ...
```

Comment (kentcdodds) [reply packages/remix-react/components.tsx:687]: I think if I were going to write this code from scratch I would do the if/else myself.

File: packages/remix-server-runtime/routeModules.ts
Comment (machour) [inline packages/remix-server-runtime/routeModules.ts:95]:
```suggestion
  [name: string]: Record<string, string> | string | string[];
```

File: packages/remix-react/components.tsx
Comment (MichaelDeBoey) [reply packages/remix-react/components.tsx:687]: > We still need keys on all elements, as we're in a loop.

@machour This part isn't in the loop, so we can omit the `key` prop

> I think if I were going to write this code from scratch I would do the if/else myself.

@kentcdodds I'll refactor it to use the separate `if`-statements

**Review by machour — APPROVED** [review]

**Comment by machour** [comment]
Still needs documentation update (and ideally tests)

File: packages/remix-react/components.tsx
Comment (kentcdodds) [inline packages/remix-react/components.tsx:678]: I think the array case should probably get into recursion for this logic rather than duplicating just the `isOpenGraphTag` logic. Maybe we should extract this logic into a function to make that possible.

File: packages/remix-react/components.tsx
Comment (kentcdodds) [inline packages/remix-react/components.tsx:693]: Let's be consistent and keep the whole thing `if/else`

**Review by kentcdodds — CHANGES_REQUESTED** [review]
This is looking great. I just have some thoughts.

File: packages/remix-react/components.tsx
Comment (kentcdodds) [reply packages/remix-react/components.tsx:678]: Based on the code I think it was already and I seem to recall this is documented somewhere.

Comment (MichaelDeBoey) [reply packages/remix-react/components.tsx:677]: I only enabled objects, not array of objects.

**Comment by MichaelDeBoey** [comment]
@machour I think it's a problem on `dev`, as this only happened after rebasing & `dev` fails too

https://github.com/remix-run/remix/runs/5580333691

**Comment by MichaelDeBoey** [comment]
@machour It seems like https://github.com/remix-run/remix/commit/32335011e2fc0b77d9ec26471f14f9a458814323 is passing, but a rebase on `dev` (https://github.com/remix-run/remix/commit/e9ef8abf8b02851425ab7486ea3cd597fe5b584f) seems to be failing 🤔

**Comment by MichaelDeBoey** [comment]
> I suspect that the problems is with the flat map part

@machour Even with the `flatMap` call, it passed a couple of commits before current `dev` HEAD

> I don't see why we're expecting an eventual array there anyways

The `value` is either `string` or `object` **or** an array of `string`s/`object`s.
To not have to do the same thing twice (once in the loop and once outside of it, like we did before), I first wrapped the `value` in an array. Since we can now have nested arrays (if the `value` was already an array), we flatten the array.
This way we're **always** having an array of `string`s/`object`s and can do the `if`/`else` only once.

**Comment by MichaelDeBoey** [comment]
@machour I first only added support for non-array objects, but @kentcdodds asked me to add support for object arrays too, so it's more consistent.

**Comment by machour** [comment]
Gotcha. I wasn't aware that arrays were supported before this PR. Thanks for baring with me 🙏

**Comment by MichaelDeBoey** [comment]
@kentcdodds @machour I also added some tests now.
I also made the types a bit stricter for `charset`, `charSet` & `title` as those were only supported anyways.

**Comment by kentcdodds** [comment]
I think we should be able to get this merged as soon as we get CI happy 👍

**Comment by MichaelDeBoey** [comment]
> get CI happy

That's where I could use some help I think 😅
I don't really see the problem in my code changes... 🤔😕

File: integration/meta-test.ts
Comment (mcansh) [inline integration/meta-test.ts:68]: this works, but seems overly verbose

```suggestion
    await expect(app.getHtml('meta[property="og:type"]')).rejects.toThrowError(
      "No element matches selector \"meta[property=\"og:type\"]\""
    );
```

**Review by mcansh — APPROVED** [review]

**Review by kentcdodds — APPROVED** [review]
This is a great change. Thank you for this PR!

File: packages/remix-react/routeModules.ts
Comment (machour) [inline packages/remix-react/routeModules.ts:75]: @MichaelDeBoey A little late to the party here, but unless I'm wrong, we shouldn't force `utf-8` here.
`windows-1252` is valid for example.

Comment (machour) [reply packages/remix-react/routeModules.ts:75]: 👍🏼  (I was looking at an older version of the HTML spec.)

**Comment by kiliman** [comment]
@kentcdodds Any chance we could look at making `meta` allow returning an array like `links`. If so, it would simplify meta handling and we wouldn't need this weird object syntax.

See PR #2497 where we discuss this. Still need to create a PR for implementing array version, but if you're open to it, I'd gladly work on it.

I think it would be nice to have similar API for both `meta` and `links`. If this is supported, even deprecate the current object syntax in favor of array of objects.
```ts
export const meta = ({data}) => [
  { title: 'my title' },
  { charset: "utf-8" },
  { property: "og:xxx", content: "url" },
  { property: "article:published_time", content: "..." },
  { name: "description", content: "..." },
  { name: "twitter:card", content: "..." },
  { property: "fb:app_id", content: "..." },
  { httpEquiv: "refresh", content: "..." },
  ...
]
```

**Comment by donavon** [comment]
I also would like to see:

*  #2497 merged
*  Array support for meta (with the current object syntax)
*  This PR #2331 reverted

I don't think we should deprecate object syntax as it's simpler. Array syntax would be your escape hatch if you need complete control.

**Comment by kentcdodds** [comment]
The benefit of the object syntax here is when we merge all the routes it can be used to override the meta values.

I'm fine with adding array syntax. I'm not sure why we need to revert #2331 though.

**Comment by kentcdodds** [comment]
As noted, the name is not meaningless, which is why I agreed and merged this PR: https://github.com/remix-run/remix/pull/2331#issuecomment-1069420261

My bad on merging it before getting Michael and Ryan's approval (I thought I had, but obvious I hadn't).

But I think it's important to consider the override case which is why I'd say we should not deprecate the object version and *only* do this with arrays in the future. I think both make sense.

**Comment by kiliman** [comment]
I was waiting for the other PR to land before adding the Array variant. I should have a PR ready soon.

**Pull request merged**
